wtf-hir/src/compiler.rs

```diff
@@ -15,7 +17,43 @@ use crate::{
 const INTERNAL_PREFIX: &str = "wtfinternal";
 const INTERNAL_SUFFIX: &str = "sdafbvaeiwcoiysxuv";
 
-pub fn compile(ast: ast::Module) -> Module {
+use std::cell::RefCell;
+use std::rc::Rc;
+
+// Thread-local error collector for simple error reporting
+thread_local! {
```

Comment:
While I appreciate this solution, this shouldn't be handled by globals but instead by using an internal struct.

So compile is still top level but compile_internal is a method and all functions it calls that need to report errors are too.